### PR TITLE
fix: review round

### PR DESCRIPTION
**Pull request by rebelchris**
Fixing issues from review round 2, in regards to context menu and sidebar.

DD-294 #done

### CODE REVIEW

**Review by rebelchris — COMMENTED** [review]
I think we can however move the useDisableFilterAlert to the sidebar in general?
And the filter page (mobile) 

Also getting an infinite loop on loading the cloudinary image?
https://res.cloudinary.com/daily-now/image/upload/f_auto/v1/placeholders/1

And side note:
The filter gets set in the database now, but never returns in the boot query

File: packages/shared/src/components/PostOptionsMenu.tsx [hunk 1]
```diff
@@ -40,9 +40,9 @@ export default function PostOptionsMenu({
   onRemovePost,
 }: PostOptionsMenuProps): ReactElement {
   useFeedSettings();
-  const { user } = useContext(AuthContext);
+  const { trackEvent } = useContext(AnalyticsContext);
   const { reportPost, hidePost } = useReportPost();
-  const { unfollowSource, blockTag } = useMutateFilters(user);
+  const { onUnfollowSource, onBlockTags } = useTagAndSource({ origin: 'feed' });
```
Comment (idoshamun) [inline packages/shared/src/components/PostOptionsMenu.tsx:45]: Let's be more explicit here just in case

```suggestion
  const { onUnfollowSource, onBlockTags } = useTagAndSource({ origin: 'post context menu' });
```

File: packages/shared/src/components/PostOptionsMenu.tsx [hunk 2]
```diff
@@ -68,34 +68,44 @@ export default function PostOptionsMenu({
...
+        '🙈 This article won’t show up on your feed anymore',
+        postIndex,
+        0,
+      );
+    }
   };
 
   const onSharePost = async () => {
-    await navigator.clipboard.writeText(post.permalink);
+    trackEvent({
```
Comment (idoshamun) [inline packages/shared/src/components/PostOptionsMenu.tsx:100]: I created a helper function for post events that you need to use to be consistent with the other post events.
I don't remember the name but look for it

File: packages/shared/src/components/PostOptionsMenu.tsx [hunk 3]
```diff
@@ -68,34 +68,44 @@ export default function PostOptionsMenu({
...
+      event_name: 'share post',
+      target_type: 'post',
+      target_id: post.id,
+      extra: JSON.stringify({ origin: 'feed' }),
```
Comment (idoshamun) [inline packages/shared/src/components/PostOptionsMenu.tsx:104]: Fix here the `feed` as above

File: packages/shared/src/components/filters/BlockedFilter.tsx [hunk 4]
```diff
@@ -1,25 +1,27 @@
...
   const [unblockItem, setUnblockItem] = useState<{
     tag?: Tag;
     source?: Source;
     action?: () => unknown;
   }>();
   const { feedSettings, isLoading } = useFeedSettings();
-  const { unblockTag, followSource } = useMutateFilters(user);
+  useDisableFilterAlert(feedSettings);
+  const { onUnblockTags, onFollowSource } = useTagAndSource({
+    origin: 'blocked-filter',
```
Comment (idoshamun) [inline packages/shared/src/components/filters/BlockedFilter.tsx:23]:
```suggestion
    origin: 'blocked filter',
```

File: packages/shared/src/components/filters/TagsFilter.tsx [hunk 5]
```diff
@@ -10,16 +10,18 @@ import useFeedSettings from '../../hooks/useFeedSettings';
...
 export default function TagsFilter(): ReactElement {
   const searchRef = useRef<HTMLDivElement>(null);
   const [query, setQuery] = useState<string>(null);
   const searchKey = getSearchTagsQueryKey(query);
   const { tagsCategories, feedSettings, isLoading } = useFeedSettings();
+  useDisableFilterAlert(feedSettings);
   const { contextSelectedTag, setContextSelectedTag, onTagContextOptions } =
     useTagContext();
-  const { onBlock } = useTag();
+  const { onBlockTags } = useTagAndSource({ origin: 'tags-filter-search' });
```
Comment (idoshamun) [inline packages/shared/src/components/filters/TagsFilter.tsx:24]:
```suggestion
  const { onBlockTags } = useTagAndSource({ origin: 'tags search' });
```

File: packages/shared/src/hooks/useDisableFilterAlert.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    if (
+      alerts?.filter &&
+      (feedSettings?.includeTags?.length ||
+        feedSettings?.blockedTags?.length ||
+        feedSettings?.excludeSources?.length ||
+        feedSettings?.advancedSettings?.length)
+    ) {
+      request(`${apiUrl}/graphql`, UPDATE_ALERTS, {
+        data: { filter: false },
+      }).then(() => setAlerts({ ...alerts, filter: false }));
```
Comment (idoshamun) [inline packages/shared/src/hooks/useDisableFilterAlert.ts:23]: This is not the right way to do it.
We can use an optimistic update for it. There are lots of examples throughout the codebase.
Optimistic update will update the UI even before the request is completed.

File: packages/shared/src/hooks/useDisableFilterAlert.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  const { alerts, setAlerts } = useAlertContext();
+
+  useEffect(() => {
+    if (
+      alerts?.filter &&
+      (feedSettings?.includeTags?.length ||
+        feedSettings?.blockedTags?.length ||
+        feedSettings?.excludeSources?.length ||
+        feedSettings?.advancedSettings?.length)
+    ) {
```
Comment (idoshamun) [inline packages/shared/src/hooks/useDisableFilterAlert.ts:20]: Does it cover the case where I already have my feed configured and I have the red dot because it's the first I see the new sidebar?

File: packages/shared/src/hooks/useMutateFilters.ts [hunk 8]
```diff
@@ -59,6 +63,12 @@ async function updateQueryData(
       });
     }),
   );
+  if (alerts.filter) {
+    await request(`${apiUrl}/graphql`, UPDATE_ALERTS, {
+      data: { filter: false },
+    });
+    setAlerts({ ...alerts, filter: false });
+  }
```
Comment (idoshamun) [inline packages/shared/src/hooks/useMutateFilters.ts:71]: Here we dispatch again the update alerts request. We should have a single place that dispatches this call and then the other places may use this function/hook.

File: packages/shared/src/hooks/useReportPost.ts [hunk 9]
```diff
@@ -17,17 +19,31 @@ type UseReportPostRet = {
...
+
   const { mutateAsync: reportPost } = useMutation<
     void,
     unknown,
     { id: string; reason: ReportReason; comment: string }
-  >((variables) =>
-    request(`${apiUrl}/graphql`, REPORT_POST_MUTATION, variables),
-  );
+  >((variables) => {
+    trackEvent({
```
Comment (idoshamun) [inline packages/shared/src/hooks/useReportPost.ts:29]: Try to use the post analytics event helper that I mentioned before here as well

File: packages/shared/src/hooks/useReportPost.ts [hunk 10]
```diff
@@ -17,17 +19,31 @@ type UseReportPostRet = {
...
+      extra: JSON.stringify({ origin: 'feed', reason: variables?.reason }),
+    });
+    return request(`${apiUrl}/graphql`, REPORT_POST_MUTATION, variables);
+  });
 
-  const { mutateAsync: hidePost } = useMutation<void, unknown, string>((id) =>
-    request(`${apiUrl}/graphql`, HIDE_POST_MUTATION, { id }),
-  );
+  const { mutateAsync: hidePost } = useMutation<void, unknown, string>((id) => {
+    trackEvent({
```
Comment (idoshamun) [inline packages/shared/src/hooks/useReportPost.ts:39]: And here

File: packages/shared/src/hooks/useTagAndSource.ts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+interface TagActionArguments {
+  tags: Array<string>;
+  category?: string;
+}
+
+interface SourceActionArguments {
+  source: Source;
+}
+
+export default function useTagAndSource({ origin }: { origin?: string }): {
```
Comment (idoshamun) [inline packages/shared/src/hooks/useTagAndSource.ts:16]: If we block source or tag through a post menu context we should add the relevant post id in the extra property

File: packages/shared/src/hooks/useTagAndSource.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+      target_type: 'tag',
+      target_id: tags[0],
+      extra: JSON.stringify({ origin }),
+    });
+    await unblockTag({ tags });
+  };
+
+  const onFollowSource = async ({ source }: SourceActionArguments) => {
+    trackEvent({
+      event_name: 'unfollow',
```
Comment (idoshamun) [inline packages/shared/src/hooks/useTagAndSource.ts:82]:
```suggestion
      event_name: 'follow',
```

File: packages/shared/src/hooks/useTagAndSource.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+      target_type: 'source',
+      target_id: source?.id,
+      extra: JSON.stringify({ origin }),
+    });
+    await followSource({ source });
+  };
+
+  const onUnfollowSource = async ({ source }: SourceActionArguments) => {
+    trackEvent({
+      event_name: 'follow',
```
Comment (idoshamun) [inline packages/shared/src/hooks/useTagAndSource.ts:92]:
```suggestion
      event_name: 'unfollow',
```

File: packages/webapp/pages/sources/[source].tsx [hunk 14]
```diff
@@ -47,7 +47,9 @@ const SourcePage = ({ source }: SourcePageProps): ReactElement => {
   );
 
   const { feedSettings } = useFeedSettings();
-  const { followSource, unfollowSource } = useMutateFilters(user);
+  const { onFollowSource, onUnfollowSource } = useTagAndSource({
+    origin: 'source-page',
```
Comment (idoshamun) [inline packages/webapp/pages/sources/[source].tsx:51]:
```suggestion
    origin: 'source page',
```

File: packages/webapp/pages/tags/[tag].tsx [hunk 15]
```diff
@@ -38,11 +38,9 @@ const TagPage = ({ tag }: TagPageProps): ReactElement => {
   const { user, showLogin } = useContext(AuthContext);
   // Must be memoized to prevent refreshing the feed
   const queryVariables = useMemo(() => ({ tag, ranking: 'TIME' }), [tag]);
-
   const { feedSettings } = useFeedSettings();
-
-  const { followTags, unfollowTags, blockTag, unblockTag } =
-    useMutateFilters(user);
+  const { onFollowTags, onUnfollowTags, onBlockTags, onUnblockTags } =
+    useTagAndSource({ origin: 'tag-page' });
```
Comment (idoshamun) [inline packages/webapp/pages/tags/[tag].tsx:43]:
```suggestion
    useTagAndSource({ origin: 'tag page' });
```

**Review by idoshamun — CHANGES_REQUESTED** [review]
Tests for the red dot are also missing.
There's still quite work to be done but we're on track 💪

File: packages/shared/src/components/PostOptionsMenu.tsx (hunk 2 above)
Comment (rebelchris) [reply packages/shared/src/components/PostOptionsMenu.tsx:100]: @idoshamun Do you mean this one?

```js
trackEvent(
      postAnalyticsEvent('share post', post, {
        extra: { media, origin: 'share bar' },
      }),
    );
```

Comment (idoshamun) [reply packages/shared/src/components/PostOptionsMenu.tsx:100]: Exactly! Another thing I noticed now when the native share is supported it should be used instead of the clipboard.
I think I have some helpers around somewhere in the article page components.

File: packages/shared/src/hooks/useDisableFilterAlert.ts (hunk 7 above)
Comment (sshanzel) [reply packages/shared/src/hooks/useDisableFilterAlert.ts:20]: This is that part. If the alerts.filter is true and it has any pre-configured values inside the `feedSettings` it will remove the red dot. This will be triggered on a load of the component where this was used.

**Comment by rebelchris** [comment]
@sshanzel @idoshamun Refactored the analytics and tag/source action props.

File: packages/shared/src/hooks/useDisableFilterAlert.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
+import { useEffect } from 'react';
+import request from 'graphql-request';
+import { UPDATE_ALERTS } from '../graphql/alerts';
+import { apiUrl } from '../lib/config';
+import { FeedSettings } from '../graphql/feedSettings';
+import useAlertContext from './useAlertContext';
+
+export default function useDisableFilterAlert(
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/hooks/useDisableFilterAlert.ts:8]: That makes sense, though I was thinking about its requirement to `feedSettings` which will make an unnecessary request at sidebar load since the data request will only be needed from the inside of each filter.

[the remark this answers is not in the document]
Comment (idoshamun) [reply packages/shared/src/hooks/useDisableFilterAlert.ts]: Now that we have a single request that fetches all the data for the sidebar that makes sense to fetch it as the sidebar opens

[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/hooks/useDisableFilterAlert.ts]: Good point. Coming right up!

**Comment by sshanzel** [comment]
I am aware of the failing test, will include the fix for it when I will be writing the test for red dot in the sidebar. Thank you! 🙏

File: packages/shared/src/components/Sidebar.spec.tsx [hunk 17]
```diff
@@ -54,6 +69,25 @@ const renderComponent = (onboardingStep = -1): RenderResult => {
   );
 };
 
+it('should remove red dot for filter alert when there is a pre-configured feedSettings', async () => {
```
Comment (idoshamun) [inline packages/shared/src/components/Sidebar.spec.tsx:72]: Where do we check for the actual call for the API?

File: packages/shared/src/components/filters/AdvancedSettings.spec.tsx [hunk 18]
```diff
@@ -137,7 +151,7 @@ it('should mutate update feed advanced settings', async () => {
       variables: { settings: params },
     },
     result: () => {
-      mutationCalled = true;
+      advancedSettingsMutation();
```
Comment (idoshamun) [inline packages/shared/src/components/filters/AdvancedSettings.spec.tsx:154]: Why did you change this mechanism?

File: packages/shared/src/components/filters/AdvancedSettings.tsx [hunk 19]
```diff
@@ -21,14 +25,28 @@ function AdvancedSettingsFilter(): ReactElement {
...
+    if (alerts?.filter) {
+      disableFilterAlert();
+    }
+
+    const enabled = !settings[id];
+    const { title } = advancedSettings.find((adv) => adv.id === id);
+
+    trackEvent({
+      event_name: `toggle ${enabled ? 'on' : 'off'}`,
+      target_type: 'advanced_settings',
```
Comment (idoshamun) [inline packages/shared/src/components/filters/AdvancedSettings.tsx:43]:
```suggestion
      target_type: 'advanced setting',
```

File: packages/shared/src/components/filters/AdvancedSettings.tsx [hunk 20]
```diff
@@ -21,14 +25,28 @@ function AdvancedSettingsFilter(): ReactElement {
...
+      target_id: `${id}_${title}`,
```
Comment (idoshamun) [inline packages/shared/src/components/filters/AdvancedSettings.tsx:44]:
```suggestion
      target_id: id.toString(),
```

File: packages/shared/src/components/filters/TagsFilter.spec.tsx [hunk 21]
```diff
@@ -141,24 +148,23 @@ it('should show login when not logged in', async () => {
...
-    expect(data).toBeTruthy();
-  });
   mockGraphQL({
     request: {
       query: ADD_FILTERS_TO_FEED_MUTATION,
       variables: { filters: { includeTags: ['webdev'] } },
     },
     result: () => {
-      mutationCalled = true;
+      addFilterMutation();
```
Comment (idoshamun) [inline packages/shared/src/components/filters/TagsFilter.spec.tsx:167]: Again why did you change this mechanism?

File: packages/shared/src/contexts/BootProvider.tsx [hunk 22]
```diff
@@ -91,12 +90,11 @@ export const BootDataProvider = ({
 
   useRefreshToken(bootData?.accessToken, refetch);
   const { user, updateUser, loadedFromCache } = useCacheUser(bootData?.user);
-  const alertContext = useAlertContext(bootData?.alerts);
   const updatedAtActive = user ? dataUpdatedAt : null;
 
   return (
     <FeaturesContextProvider flags={bootData?.flags}>
-      <AlertContext.Provider value={alertContext}>
+      <AlertContextProvider alerts={bootData?.alerts}>
```
Comment (idoshamun) [inline packages/shared/src/contexts/BootProvider.tsx:97]: This should be a child of the `AuthContext`

File: packages/shared/src/hooks/useTagAndSource.ts [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+interface TagActionArguments {
+  tags: Array<string>;
+  category?: string;
+}
+
+interface SourceActionArguments {
+  source: Source;
+}
+
+export default function useTagAndSource({
```
Comment (idoshamun) [inline packages/shared/src/hooks/useTagAndSource.ts:17]: All the subfunctions here are missing showing login in case the user is not signed in except the follow tags.
We can use the origin prop as the first argument for the `showLogin` func.

Let's also provide the `postId` on every tracked event here, just in case. Btw, our convention in the analytics system is an underscore case. So please change to `post_id: postId`

**Review by idoshamun — CHANGES_REQUESTED** [review]
Wow almost there!

File: packages/shared/src/hooks/useTagAndSource.ts (hunk 23 above)
Comment (rebelchris) [reply packages/shared/src/hooks/useTagAndSource.ts:17]: Ah didn't add the call as it's virtually impossible to trigger these without being logged in.
But a good fallback! will do 🤘

File: packages/shared/src/components/Sidebar.spec.tsx (hunk 17 above)
Comment (sshanzel) [reply packages/shared/src/components/Sidebar.spec.tsx:72]: I actually tried initially, but for some reason even when the `alerts` data was updated already, the mock doesn't do the rerender for the components making the `useTagsAndSources` use the old data. I can't somehow make it possible with the test case. Though locally, it works as intended, so I mocked whether the function will be called.

File: packages/shared/src/components/filters/AdvancedSettings.spec.tsx (hunk 18 above)
Comment (sshanzel) [reply packages/shared/src/components/filters/AdvancedSettings.spec.tsx:154]: I initially was trying to contain the condition in a single jest mock to see if it will be called twice, but then I decided to separate them for clarity, then I just kept it to be consistent with each other.

File: packages/shared/src/components/filters/TagsFilter.spec.tsx (hunk 21 above)
Comment (sshanzel) [reply packages/shared/src/components/filters/TagsFilter.spec.tsx:167]: I just kept it to be consistent with each other but we can revert it if necessary.

File: packages/shared/src/components/filters/AdvancedSettings.spec.tsx (hunk 18 above)
Comment (idoshamun) [reply packages/shared/src/components/filters/AdvancedSettings.spec.tsx:154]: I find this new way less readable. Can you revert please everywhere?

File: packages/shared/src/components/Sidebar.spec.tsx (hunk 17 above)
Comment (idoshamun) [reply packages/shared/src/components/Sidebar.spec.tsx:72]: Ok that's tricky but let's keep it this way for now

File: packages/shared/src/components/filters/AdvancedSettings.spec.tsx (hunk 18 above)
Comment (sshanzel) [reply packages/shared/src/components/filters/AdvancedSettings.spec.tsx:154]: Sure thing.

**Review by idoshamun — APPROVED** [review]
Woohooo!

**Comment by sshanzel** [comment]
OH YEAHHHH 🚢
